polynote-frontend/polynote/ui/component/comment.ts

```diff
@@ -0,0 +1,517 @@
+"use strict";
+
+
+import {button, Content, div, span, tag, TagElement, textarea, textbox} from "../util/tags";
+import { CreateComment, CurrentIdentity, DeleteComment, UIMessageRequest, UIMessageTarget, UpdateComment } from "../util/ui_event";
+import * as uuidv4 from 'uuid/v4';
+import {CellComment} from "../../data/data";
+import {PosRange} from "../../data/result";
+import * as monaco from "monaco-editor";
+import {editor, IDisposable} from "monaco-editor";
+import IStandaloneCodeEditor = editor.IStandaloneCodeEditor;
+import IModelContentChangedEvent = editor.IModelContentChangedEvent;
+import TrackedRangeStickiness = editor.TrackedRangeStickiness;
+
+export type CommentID = string
+
+// to ensure UUID uniqueness, make sure to only create new comments with createCellComment!
+export function createCellComment({range, author, createdAt, content}: Omit<CellComment, 'uuid'>): CellComment {
+    const uuid = uuidv4();
+    return new CellComment(
+        uuid,
+        range,
+        author,
+        createdAt,
+        content,
+    )
+}
+
+// Comment Handler is in charge of comments for a particular cell
+export class CommentHandler extends UIMessageTarget {
+    private comments: Record<CommentID, CellComment> = {}; // holds all comments (root and replies)
+    private commentUIs: Record<string, CommentUI> = {};    // lookup table for comment UIs (keys are ranges, serialized as strings)
+    private commentRoots: Record<string, CommentID> = {};  // lookup table for range (serialized as string) to root comments
+    private highlights: Record<string, string[]> = {};     // deltaDecorations, range string -> monaco decoration ids.
+
+    constructor(readonly cellId: number, readonly editor: IStandaloneCodeEditor) {
+        super();
+
+        this.subscribe(CreateComment, (cellId, comment) => {
+            this._add(comment);
+        });
+        this.subscribe(UpdateComment, (cellId, commentId, range, content) => {
+            this._update(commentId, range, content);
+        });
+        this.subscribe(DeleteComment, (cellId, commentId) => {
+            this._delete(commentId);
+        });
```

Comment:
`add`, `update` and `delete` have private `_add`, `_update` and `_delete` methods in order to apply external changes (from somewhere else, update internal state and the UI too) and internal changes (from the UI, so we only need to update internal state).

Reply:
Also, note that these events will get bubbled up to the notebook which is responsible for sending messages to the backend.